meadow-client/src/Blockly.js

```diff
@@ -0,0 +1,92 @@
+/*eslint-env node*/
+'use strict';
+
+exports.createBlocklyInstance_ = function() {
+    return require('node-blockly/browser');
+}
+
+exports.createWorkspace_ = function(blockly, workspaceDiv, toolboxDiv) {
+    var workspace = blockly.inject(workspaceDiv, {toolbox: document.getElementById(toolboxDiv)
+                                                 , collapse : true 
+                                                 , comments : true 
+                                                 , disable : true
+                                                 , maxBlocks : Infinity
+                                                 , trashcan : true
+                                                 , horizontalLayout : false
+                                                 , toolboxPosition : 'start'
+                                                 , css : true
+                                                 , media : 'https://blockly-demo.appspot.com/static/media/'
+                                                 , rtl : false
+                                                 , scrollbars : true
+                                                 , sounds : true
+                                                 , oneBasedIndex : true
+                                                 , zoom:
+                                                   {controls: true,
+                                                    wheel: false,
+                                                    startScale: 1.0,
+                                                    maxScale: 3,
+                                                    minScale: 0.3,
+                                                    scaleSpeed: 1.2}
+                                                 , grid:
+                                                    {spacing: 20,
+                                                     length: 3,
+                                                     colour: '#ccc',
+                                                     snap: true}
+                                                 });
+    blockly.svgResize(workspace);
+    return workspace;
+
+}
+
+exports.resizeBlockly_ = function(blocklyState) {
+    var Blockly = blocklyState.blockly;
+    var workspace = blocklyState.workspace;
+    Blockly.svgResize(workspace);
+    workspace.render();
+}
+
+function removeUndefinedFields(obj) {
+    for (var propName in obj) { 
+        if (obj[propName] === undefined) {
+          delete obj[propName];
+        }
+      }
+  }
+
+function removeEmptyArrayFields(obj) {
+    for (var propName in obj) { 
+        if (Array.isArray(obj[propName]) && obj[propName].length == 0) {
+          delete obj[propName];
+        }
+      }
+}
+
+exports.addBlockType_ = function(blocklyState, name, block) {
+    var Blockly = blocklyState.blockly;
+    removeUndefinedFields(block);
+    removeEmptyArrayFields(block);
```

Comment:
Hmm...there's something wrong here. This function changes `block` but you wouldn't know it from the PureScript type signature. That's going to be a source of hard-to-trace bugs.

🤔 

Hate to say it, but you might need to move these mutable things into something like the `ST` monad.

Reply:
It's an `Effect Unit`, this tells me something but I see what you mean, I couldn't use that `Block` again. What I wanted was an immutable object but since PS uses ECMA 5 and not 6 it seems difficult (difficult to navigate the web to find out how)

Reply:
ok, this is an easy clone using a nasty hack (obviously) so now we avoid secretly mutating the input object. `BlocklyState` is always going to be mutated, that is the point of it existing and blockly can't work without doing this.

Reply:
The `ST` monad would let you express, "this variable is mutable," at the type level. That's its job. Up to you if you want to convert it...

Reply:
ok, I'm using `ST` everywhere mutation takes place, needs review though